### ISSUE TITLE
Will the editor soon be usable for GTA San Andreas the definitive edition?

### ISSUE THREAD

**Issue by EHSBprogrammer** [gtasa-savegame-editor/gtasa-savegame-editor#179]
I have the Rockstar GTA definitive edition trilogy, the trainer does not work with that version

**Comment by lfuelling** [comment]
Hi, 
could you send me a savefile?
The progress is not really important, as I need to add compatibility with the new file format (if anything changed).

**Comment by EHSBprogrammer** [comment]
Yes, here it is

 

Van: Lerk ***@***.***> 
Verzonden: zaterdag 17 juni 2023 17:58
Aan: gtasa-savegame-editor/gtasa-savegame-editor ***@***.***>
CC: EHSBprogrammer ***@***.***>; Author ***@***.***>
Onderwerp: Re: [gtasa-savegame-editor/gtasa-savegame-editor] Add compatibility with GTA San Andreas definitive edition (Issue #179)

 

Hi,
could you send me a savefile?
The progress is not really important, as I need to add compatibility with the new file format (if anything changed).

**Comment by lfuelling** [comment]
Sorry, attachments don't seem to work with GitHub issue comment mails. 
Could you please attach it to a comment using the web ui?

**Comment by EHSBprogrammer** [comment]
[GTASAsf1.zip](https://github.com/gtasa-savegame-editor/gtasa-savegame-editor/files/11855929/GTASAsf1.zip)

I hope you have the file now

**Comment by EHSBprogrammer** [comment]
Have my files been helpfull? has the trainer been updated  for my gta version yet?

**Comment by EHSBprogrammer** [comment]
I mean the savegame editor, i'm a little mixed up between trainers and savegame editors

**Comment by lfuelling** [comment]
Unfortunately it seems like that edition of GTA:SA is using a different savefile format.
All the supported versions until the definitive edition used a savegame format with the extension `.b`, where the contained data was split into blocks using the literal ASCII text `BLOCK` as a marker. ([our documentation about that format](https://gtasa-savegame-editor.github.io/docs/#/))

The savegame you've attached has the extension `.sav` and parsing with the savegame CLI fails completely because the `BLOCK` markers are not in there anymore.
Dumping the file next to the 100% savegame (using `hexdump -C <file>`) reveals that the new savefile format is very, if not completely, different from the current one, so adding support for it will not be that easy. :(

**Comment by xxeevv** [comment]
uh... it's a pity((

**Comment by lfuelling** [comment]
I can certainly help with implementation if someone manages to analyze the new format, but I am not very good at this. 
I originally took maintainership to fix UI issues and add better compatibility with macOS and GTA:SA for macOS. 🙈

**Comment by Apocalypse02** [comment]
You can just rename to old format and it will be supported then you can rename it back for game to load it, it's how I did it before.

**Comment by lfuelling** [comment]
> You can just rename to old format and it will be supported then you can rename it back for game to load it, it's how I did it before.

You mean that you changed `.sav` to `.b` and that made it work?
Are you sure that editing anything in the save and then saving it didn't corrupt it?

As far as I came with analysis the format is entirely different and neither editing nor loading an edited save should work. But I also don't have the Definitive Edition so I can't test it out :(

**Comment by Apocalypse02** [comment]
> > You can just rename to old format and it will be supported then you can rename it back for game to load it, it's how I did it before.
> 
> You mean that you changed `.sav` to `.b` and that made it work? Are you sure that editing anything in the save and then saving it didn't corrupt it?
> 
> As far as I came with analysis the format is entirely different and neither editing nor loading an edited save should work. But I also don't have the Definitive Edition so I can't test it out :(
 
I did exactly that, well I can just tell you you should avoid torrent sites and Dodi repacks those things can be dangerous if you don't watch out what you are downloading.

**Comment by EHSBprogrammer** [comment]
> Hi, could you send me a savefile? The progress is not really important, as I need to add compatibility with the new file format (if anything changed).

My save file i hope it will help you, the editor asks for savegame file and gamefile and int does not accept both
[GTASAsf1.zip](https://github.com/gtasa-savegame-editor/gtasa-savegame-editor/files/13254984/GTASAsf1.zip)

**Comment by EHSBprogrammer** [comment]
Good evening,

I saw in the reactions that with renaming I should be able to edit my GTA San Andreas Definitive Edition files with the save game editor.
If I try to start the dowloaded Windows file I can not start the executable because it needs Java Runtime Environment 11.
I tried to dowload the file that is linked bij the editor but it still does not work. What should I do to make it work? Could anyone help me please?

Thanks!

**Comment by EHSBprogrammer** [comment]
I just saw that there is a separate isue in github for this, i'll read that.

**Comment by EHSBprogrammer** [comment]
Dowloading Java JDK from Oracle made the editor start but renaming the savegamefile from .save to .b does not work at all.
I hope this will be fixed someday in the future.

**Comment by SichelBichel** [comment]
Is the DE-support Project still active for the Save Editor? i could provide some resources, as i am interested in such a Editor as well. DE Modding is almost non existent (besides some Skin mods)
i do also have a few 100 hours of Unreal Engine experience which might help

**Comment by lfuelling** [comment]
Sure, if you can contribute anything please feel free to do so.
I currently just don't have the time to analyze the game (which I also would need to buy in the first place, I only have the original GTA SA on macOS and Windows), but I'd be happy to review and merge any contribution towards that goal.

You might find some hints (if the save file format is similar) in the existing docs: 
https://gtasa-savegame-editor.github.io/docs/#/

**Comment by SichelBichel** [comment]
> You might find some hints (if the save file format is similar) in the existing docs:

Yeah thats what im currently trying to achieve. those .SAV files created by DE are nowhere near humanly readable. the issue is that they differ from any known UE4 .SAV formats. Rockstar Probably put some sort of encryption on them. Every single tool that can read SAVs crashes or throws errors upon opening a DE Save until now. i'll see if i can find a way to take a look at them (apart from plain HEX editing)

**Comment by lfuelling** [comment]
You could also try to extract them in a way similar to the savegame editor [command line tool](https://github.com/gtasa-savegame-editor/gtasa-savegame-editor/tree/main/cli).
In the (old/previous) savegame format, the blocks mentioned in the docs are separated by literal `BLOCK` markers, maybe the marker is now some hex code instead of ascii?

**Comment by SkipperSkipTR** [comment]
I don't know if you are still working on this but I found some .dat files named GTASASaveInfo in the game files. They seem to be zlib compressed with a header. I don't know if they are useful but they look like this after trying to decompress them:

![Image](https://github.com/user-attachments/assets/656dc4a9-eb5c-4d65-8042-777a05716a3c)

You can find the files here: https://drive.google.com/file/d/1Kn7pIWeuLdID1H3nIgRZ_PbdpA1yF8JC/view?usp=sharing

**Comment by SichelBichel** [comment]
Thanks! i am rather busy with other stuff at the moment. However i will try to reach out to the folks from https://gtasnp.com/ as they seem to have figured out something, since they have a working Save-Reader on there

**Comment by SichelBichel** [comment]
Turns out the creator of gtasnp is here on Github xD
@Samutz Sorry for the unsolicited tag, but i thought you might have some insight that could help us here.
The DE Save-reader from gtasnp.com seems to work vrey well with the new .SAVs, and we're trying to figure out how.
Any pointers would be much appreciated!

**Comment by Samutz** [comment]
Most of what I've found I documented in this post: https://gtaforums.com/topic/893779-san-andreas-save-file-companion/page/4/#comment-1071725494
And then this post further down has some details on the checksum, included changes made to it in last December's update: https://gtaforums.com/topic/893779-san-andreas-save-file-companion/page/4/#comment-1071725494

I hope that helps. Let me know if you have questions on any specifics.

The GTASASaveInfo being zlib compressed is news to me. Since the saves contain a copy of it in their METADATA block, I had suspected that the game uses it to handle changes made between versions. For example, VC DE has added global variables a few times over the past several updates, causing their offsets to shift within the SCRIPTS block. I will definitely look in to that some more when I have the time.
